proposals/0002.md

```diff
@@ -0,0 +1,272 @@
+Usually it's not very useful, because it reduces flexibility, but doesn't simplify code much.
+
+## Root cause errors
+
+These are errors that don't wrap other errors inside.
+They may contain multiple structures and a lot of details, but their `source` always returns `None`.
+A root error can be just a variant of a bigger error with other variants, some of which may be even
+wrappers.
+
+### Statically typed root cause error
+
+The error has a statically defined meaning and details.
+It's easy to track it by humans and analyze it in the code.
+It requires scrutiny and planning to design, but it's easy to reuse in different parts of the code.
+
+```rust
+#[derive(Debug, thiserror::Error)]
+pub enum FileLoadError {
+    ...
+    #[error("the file contains invalid value {0}")
+    InvalidValue(u32),
+    ...
+}
+```
+
+### Textual root cause error
+
+The error is defined only as a text.
+It may be harder to track by humans it and it's almost impossible to analyze in the code.
+It's very easy to create new error variants, but they aren't easily reusable.
+
+Textual errors are often implemented as `String`s, but it's a subpar solution.
+It's much better to use `anyhow::Error`, because unlike `String` it implements
+`std::error::Error` making its processing easier and more consistent.
+The `anyhow` crate also provides the `anyhow!` macro, which works just like `format!`.
+
+```rust
+fn check_value(value: u32) -> Result<(), anyhow::Error> {
+    ...
+    .ok_or_else(|| anyhow::anyhow!("the file contains an unexpected value {}", value))?;
+    ...
+}
+```
+
+## Wrapper errors
+
+These are errors that wrap other errors inside. Their `source` always returns the inner error.
+A wrapper error can be just a variant of a bigger error with other variants, some of which may be
+even roots.
+
+### Statically typed wrapper error
+
+The error has a statically defined cause and details.
+It's easy to track it by humans and analyze it in the code.
+It requires scrutiny and planning to design, but it's easy to reuse in different parts of the code.
+
+```rust
+#[derive(Debug, thiserror::Error)]
+pub enum FileLoadError {
+    ...
+    #[error("failed to open the file '{path}'")]
+    FileSystemError {
+        path: String,
+        #[source]
+        source: std::io::Error,
+    },
+    ...
+}
+```
+
+Sometimes it's necessary to wrap an error not as a source, but as another case of the wrapper.
+It's supported by `thiserror` in a form of the `transparent` annotation.
+All the `std::error::Error` and `Display` calls are then passed directly to the internal error
+effectively hiding the wrapper.
+
+```rust
+#[derive(Debug, thiserror::Error)]
+pub enum FileLoadError {
+    ...
+    #[error(transparent)]
+    Other(#[from] anyhow::Error)
+    ...
+}
+```
+
+### Statically typed wrapper error when source has multiple types
+
+The error with multiple causes expressed as different types.
```

Comment:
I'm unsure what you're trying to express here. Maybe we could work on it together? Could you explain it in different words to me?

Reply:
You're right, it's kind of difficulty to decode when reading with a fresh mind. I've expanded it, is it better now?

Reply:
I'm still not quite sure I understand the issue :/ Something that might help me, at least, is a before and after. So what one might think is possible (the before) and what we actually have to do (the after - which I think is the example you've already provided). Sorry if this is me just being slow to the uptake :smile:

Reply:
I've added a "before" example, with a dumb, but ugly workaround. Thank you for a solution! Is it clearer now?